src/test/scala/firrtlTests/MemoryInitSpec.scala

```diff
@@ -213,13 +214,127 @@ class MemInitSpec extends FirrtlFlatSpec {
+       |    m.r.en <= rEnable
+       |    rData <= m.r.data
+       |
+       |    m.w.clk <= clock
+       |    m.w.addr <= wAddr
+       |    m.w.en <= wEnable
+       |    m.w.data <= wData
+       |    m.w.mask is invalid
+       |
+       |  module Top:
+       |    input clock : Clock
+       |    input rAddr : UInt<5>
+       |    input rEnable : UInt<1>
+       |    input wAddr : UInt<5>
+       |    input wData : UInt<8>
+       |    input wEnable : UInt<1>
+       |    output rData : UInt<8>[2]
+       |
+       |    inst c1 of Child
+       |    c1.clock <= clock
+       |    c1.rAddr <= rAddr
+       |    c1.rEnable <= rEnable
+       |    c1.wAddr <= wAddr
+       |    c1.wData <= wData
+       |    c1.wEnable <= wEnable
+       |
+       |    inst c2 of Child
+       |    c2.clock <= clock
+       |    c2.rAddr <= rAddr
+       |    c2.rEnable <= rEnable
+       |    c2.wAddr <= wAddr
+       |    c2.wData <= wData
+       |    c2.wEnable <= wEnable
+       |
+       |    rData[0] <= c1.rData
+       |    rData[1] <= c2.rData
+       |""".stripMargin
+
+  // Absolute references to the memory objects in each child module
+  val child1MRef = CircuitTarget("Top").module("Top").instOf("c1", "Child").ref("m")
+  val child2MRef = CircuitTarget("Top").module("Top").instOf("c2", "Child").ref("m")
+  // Final deduplicated reference
+  val dedupedRef = CircuitTarget("Top").module("Child").ref("m")
+
+  "MemoryRandomInitAnnotation" should "randomize memory in single deduped module" in {
+    val annos = Seq(
+      MemoryRandomInitAnnotation(child1MRef),
+      MemoryRandomInitAnnotation(child2MRef)
+    )
+    val result = compile(dedupTest, annos)
+    result should containLine("      m[initvar] = _RAND_0[7:0];")
+  }
+
+  "MemoryScalarInitAnnotation" should "initialize memory to 0 in deduped module" in {
+    val annos = Seq(
+      MemoryScalarInitAnnotation(child1MRef, value = 0),
+      MemoryScalarInitAnnotation(child2MRef, value = 0)
+    )
+    val result = compile(dedupTest, annos)
+    result should containLine("      m[initvar] = 0;")
+  }
+
+  "MemoryArrayInitAnnotation" should "initialize memory with array of values in deduped module" in {
+    val values = Seq.tabulate(32)(ii => 2 * ii + 5).map(BigInt(_))
+    val annos = Seq(
+      MemoryArrayInitAnnotation(child1MRef, values),
+      MemoryArrayInitAnnotation(child2MRef, values)
+    )
+    val result = compile(dedupTest, annos)
+
+    values.zipWithIndex.foreach {
+      case (value, addr) =>
+        result should containLine(s"      m[$addr] = $value;")
+    }
+  }
+
+  "MemoryFileInlineAnnotation" should "emit $readmemh in deduped module" in {
+    val annos = Seq(
+      MemoryFileInlineAnnotation(child1MRef, filename = "text.hex"),
+      MemoryFileInlineAnnotation(child2MRef, filename = "text.hex")
+    )
+    val result = compile(dedupTest, annos)
+    result should containLine("""$readmemh("text.hex", """ + dedupedRef.name + """);""")
+  }
 }
```

Comment:
I'd like to see a negative test for when not all of the instances have the annotation and the annotations thus don't dedup. I don't think you need it for every single annotation type, but for at least one of them.